### PR TITLE
Add catch safeguard for `preloadImage`

### LINKED ISSUE #9326: Uncaught promise rejections when using `preloadImage`
## Bug Description
There's a console error displayed when saving a new page template with images.
Let's figure out if there's actually anything broken and add handle the error properly.

## Steps to Reproduce
1. Create a new page with some images
2. Go to the Page Templates tab in the library and add click to save a new template
3. See the console error

## Screenshots
Screenshot 2021-10-07 at 12 31 20
Screenshot 2021-10-07 at 12 30 50

## Additional Context

- Plugin Version:
- WordPress Version:
- Operating System:
- Browser:

### PR DESCRIPTION
**Pull request by amovar18**
## Context

This PR aims to add safeguard for `preLoadImage` to catch uncaught promise rejections.

## Testing Instructions

- [x] This is a non-user-facing change and requires no QA

This PR can be tested by following these steps:

1.

## Checklist

- [ ] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [x] I have tested this code to the best of my abilities
- [x] I have verified accessibility to the best of my abilities ([docs](https://github.com/google/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [x] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [x] This code is covered by automated tests (unit, integration, and/or e2e) to verify it works as intended ([docs](https://github.com/google/web-stories-wp/tree/main/docs#testing))
- [x] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

Fixes #9326

### CODE REVIEW

File: packages/media/src/getImageDimensions.js [hunk 1]
```diff
@@ -26,7 +26,7 @@ import preloadImage from './preloadImage';
  * @return {Promise} Image dimensions object.
  */
 const getImageDimensions = async (src) => {
-  const img = await preloadImage(src);
+  const img = await preloadImage(src).catch(() => {});
```
Comment (miina) [inline packages/media/src/getImageDimensions.js]: `preloadImage` is used with `await` here everywhere, would adding `catch` like this do anything?
Probably should use `try` and `catch` instead.

File: packages/media/src/getImageDimensions.js [hunk 2]
```diff
@@ -26,7 +26,12 @@ import preloadImage from './preloadImage';
...
 const getImageDimensions = async (src) => {
-  const img = await preloadImage(src).catch(() => {});
+  let img;
+  try {
+    img = await preloadImage(src);
+  } catch (error) {
+    //ignore
+  }
   return {
     width: img.naturalWidth,
```
Comment (miina) [inline packages/media/src/getImageDimensions.js]: Now this would give an error if the `img` is `null` / `undefined`, no?

File: packages/media/src/getImageDimensions.js [hunk 3]
```diff
@@ -26,7 +26,12 @@ import preloadImage from './preloadImage';
  * @return {Promise} Image dimensions object.
  */
 const getImageDimensions = async (src) => {
-  const img = await preloadImage(src).catch(() => {});
+  let img;
+  try {
+    img = await preloadImage(src);
+  } catch (error) {
+    //ignore
```
Comment (miina) [inline packages/media/src/getImageDimensions.js]: Here and elsewhere:
```suggestion
    // Ignore.
```

File: packages/story-editor/src/elements/media/imageDisplay.js [hunk 4]
```diff
@@ -83,7 +83,12 @@ function ImageDisplay({ element, box, previewMode }) {
     if (resourceList.get(resource.id)?.type !== 'fullsize' && resource.src) {
       timeout = setTimeout(async () => {
         const url = getProxiedUrl(resource, resource.src);
-        const preloadedImg = await preloadImage(url, srcSet).catch(() => {});
+        let preloadedImg;
+        try {
+          preloadedImg = await preloadImage(url, srcSet).catch(() => {});
```
Comment (miina) [inline packages/story-editor/src/elements/media/imageDisplay.js]: This still has a `catch` here.

Also, if `preloadedImage` is `null` / `undefined`, it will still error below with `preloadedImg.currentSrc`.

File: packages/story-editor/src/utils/getBlurHashFromImage.js [hunk 5]
```diff
@@ -37,7 +37,12 @@ const getImageData = (image) => {
...
 
 const getBlurHashFromImage = async (src) => {
-  const image = await preloadImage(src).catch(() => {});
+  let image;
+  try {
+    image = await preloadImage(src);
+  } catch (error) {
+    //ignore
+  }
   const imageData = getImageData(image);
```
Comment (miina) [inline packages/story-editor/src/utils/getBlurHashFromImage.js:46]: Wouldn't this also throw an error when passing `null` / `undefined` as `image`?

Would be great if you could actually test out if these cases work as expected when you purposefully trigger an error in `preloadImage`.

**Review by miina — CHANGES_REQUESTED** [review]

File: packages/media/src/getImageDimensions.js [hunk 6]
```diff
@@ -26,7 +26,12 @@ import preloadImage from './preloadImage';
...
  */
 const getImageDimensions = async (src) => {
-  const img = await preloadImage(src);
+  let img;
+  try {
+    img = await preloadImage(src);
+  } catch (error) {
+    //ignore
+  }
   return {
```
Comment (swissspidy) [inline packages/media/src/getImageDimensions.js]: If running into the `catch` above, `img` will be `undefined` still, and trying to access `img.naturalWidth` or `img.naturalHeight` will break.

Second, no need to do `catch (errror) {` if `error` is unused. Can just do ` catch {`

I'd probably write the whole thing like this:

```js
try {
    const img = await preloadImage(src);
    return {
      width: img.naturalWidth,
      height: img.naturalHeight,
    };
  } catch {
    return { width: 0, height: 0 };
  }
```

But I am skeptical that we actually need the `catch()` here.

Everywhere `getImageDimensions` is being used there already seem to be some `catch` in place.

Adding one here and returning `width:0, height: 0` would unexpectedly break things in other ways.

**So in short:** I don't think we should add a `catch` here at all.

File: packages/story-editor/src/elements/media/imageDisplay.js [hunk 7]
```diff
@@ -83,7 +83,12 @@ function ImageDisplay({ element, box, previewMode }) {
     if (resourceList.get(resource.id)?.type !== 'fullsize' && resource.src) {
       timeout = setTimeout(async () => {
         const url = getProxiedUrl(resource, resource.src);
-        const preloadedImg = await preloadImage(url, srcSet);
+        let preloadedImg;
+        try {
+          preloadedImg = await preloadImage(url, srcSet);
+        } catch (error) {
+          //ignore
```
Comment (swissspidy) [inline packages/story-editor/src/elements/media/imageDisplay.js]: There should be a `return` here.

If you just `catch`, accessing `preloadedImg.currentSrc` below will cause an error because `preloadedImg` is `undefined`.

Also, can just do `catch {`, without `erorr`.

File: packages/story-editor/src/utils/getBlurHashFromImage.js (hunk 5 above)
Comment (amovar18) [reply packages/story-editor/src/utils/getBlurHashFromImage.js:46]: @miina this wont cause a problem because it is being used at only one place that it in `useDetectBlurHash` which already wraps `getPreloadImage` in a try catch thinking of it now I think should this even be wrapped in try catch since useDetectBlurHash wraps it as a whole?

File: packages/story-editor/src/elements/media/imageDisplay.js (hunk 4 above)
Comment (amovar18) [reply packages/story-editor/src/elements/media/imageDisplay.js]: moved the whole if condition in try catch

File: packages/media/src/getImageDimensions.js (hunk 2 above)
Comment (amovar18) [reply packages/media/src/getImageDimensions.js]: https://github.com/google/web-stories-wp/pull/10068#discussion_r778719305

File: packages/story-editor/src/utils/getBlurHashFromImage.js (hunk 5 above)
Comment (miina) [reply packages/story-editor/src/utils/getBlurHashFromImage.js:46]: Since this is an util that could in theory be used elsewhere as well, then to be on the safe side, I think it's still better to use `try` here as well, not assuming the util will only ever be used in one place. Since the try-catch around using this util inside `useDetectBlurHash` is used for `getBlurHashFromImage` only (as `saveBlurHash` has its own error handling), might make more sense to remove it from there instead but still keep it in the util.

Comment (amovar18) [reply packages/story-editor/src/utils/getBlurHashFromImage.js:46]: Leaving getBlurhashFromImage with a try/catch

Comment (miina) [reply packages/story-editor/src/utils/getBlurHashFromImage.js:46]: Ok, but let's make sure `getImageData` doesn't trigger error here then as well if `preloadImage` fails.

File: packages/story-editor/src/utils/getBlurHashFromImage.js [hunk 8]
```diff
@@ -37,7 +37,12 @@ const getImageData = (image) => {
 };
 
 const getBlurHashFromImage = async (src) => {
-  const image = await preloadImage(src);
+  let image;
+  try {
+    image = await preloadImage(src);
+  } catch {
+    // Ignore
```
Comment (spacedmonkey) [inline packages/story-editor/src/utils/getBlurHashFromImage.js]: We should not ignore this. If the image fails to load here, we should error out or something. 

```suggestion
    return Promise.reject(new Error('image failed to load'));
```

**Comment by spacedmonkey** [comment]
What about this function? 

https://github.com/google/web-stories-wp/blob/636aa1e447226936c058d40d7983c647fe4aead8/packages/media/src/getImageDimensions.js#L28-L34

**Comment by amovar18** [comment]
> What about this function?
> 
> https://github.com/google/web-stories-wp/blob/636aa1e447226936c058d40d7983c647fe4aead8/packages/media/src/getImageDimensions.js#L28-L34
@spacedmonkey 
https://github.com/google/web-stories-wp/pull/10068#discussion_r778719305

File: packages/story-editor/src/utils/getBlurHashFromImage.js (hunk 5 above)
Comment (miina) [reply packages/story-editor/src/utils/getBlurHashFromImage.js:46]: (Marking it back to unresolved since currently, `getImageData` would still error if we pass `undefined` image there.)

Comment (amovar18) [reply packages/story-editor/src/utils/getBlurHashFromImage.js:46]: It will now return a `rejected` Promise if getPreloadImage fails to load the image

File: packages/story-editor/src/elements/media/imageDisplay.js [hunk 9]
```diff
@@ -83,13 +83,18 @@ function ImageDisplay({ element, box, previewMode }) {
...
-        const preloadedImg = await preloadImage(url, srcSet);
-        if (mounted) {
-          resourceList.set(resource.id, {
-            type: 'fullsize',
-          });
-          setSrc(preloadedImg.currentSrc);
-          setSrcType('fullsize');
+        let preloadedImg;
+        try {
+          preloadedImg = await preloadImage(url, srcSet);
```
Comment (swissspidy) [inline packages/story-editor/src/elements/media/imageDisplay.js]:
```suggestion
        try {
          const preloadedImg = await preloadImage(url, srcSet);
```

**Review by miina — APPROVED** [review]
LGTM but would be good to get an approval from @spacedmonkey / @swissspidy as well as your team has mostly worked on the media part.

**Review by swissspidy — APPROVED** [review]
